BoxLEDs.cpp

```diff
@@ -12,21 +12,59 @@ void BoxLEDs::begin() {
     _stateGreen = LED_PWM_MIN;
     _stateBlue = LED_PWM_MIN;
 
-    _rainbowStepState = 0;
-
-    setIdleType(IDLE_TYPE::RAINBOW);
+    _activeAnimationRunning = false;
+    setIdleAnimation(ANIMATION_TYPE::PULSE, ANIMATION_COLOR::TEAL);
+    //setActiveAnimation(ANIMATION_TYPE::BLINK, ANIMATION_COLOR::YELLOW, 10000);
 }
 
 void BoxLEDs::loop() {
-    if (_idleType == IDLE_TYPE::RAINBOW) {
-        setAll(_wheel(_rainbowStepState));
-        if (_rainbowStepState < 255) {
-            _rainbowStepState++;
+    if (_activeAnimationRunning == true && _timer.isRunning()) {
+        _handleAnimation(&_activeAnimation);
+    } else {
+        _idleAnimation.state++;
+        _handleAnimation(&_idleAnimation);
+    }
+
+}
+
+void BoxLEDs::_handleAnimation(ANIMATION* animation) {
+    setIntervalForAnimationType(animation->type);
+    if (animation->type == ANIMATION_TYPE::RAINBOW) {
+        setAll(_wheel(animation->state));
+        if (animation->state < 255) {
+            animation->state = animation->state++;
         } else {
-            _rainbowStepState = 0;
+            animation->state = 0;
         }
-    } else if (_idleType == IDLE_TYPE::PARTY) {
+    } else if (animation->type == ANIMATION_TYPE::PARTY) {
         setAll(_wheel(random(255)));
+    } else if (animation->type == ANIMATION_TYPE::PULSE) {
+        if (animation->direction == ANIMATION_DIRECTION::UP) {
+            if (animation->state <= LED_PWM_MAX - animation->step) {
+                animation->state = animation->state+animation->step;
+            } else {
+                animation->direction = ANIMATION_DIRECTION::DOWN;
+                animation->state = LED_PWM_MAX;
+            }
+        } else { 
+            if (animation->state >= LED_PWM_MIN + animation->step) {
+                animation->state -= animation->step;
+            } else {
+                animation->direction = ANIMATION_DIRECTION::UP;
+                animation->state = LED_PWM_MIN;
+            }
+        }
+        setAll(_transformPulse(animation->state, animation->color));
```

Comment:
Please ignore LED_PWM_MIN/MAX should only be handled within the setLED methods (as its just a workaround).

```suggestion
    } else if (animation->type == ANIMATION_TYPE::PULSE) {
        if (animation->direction == ANIMATION_DIRECTION::UP) {
            if (animation->state <= animation->step) {
                animation->state = animation->state+animation->step;
            } else {
                animation->direction = ANIMATION_DIRECTION::DOWN;
                animation->state = LED_PWM_MAX;
            }
        } else { 
            if (animation->state >= animation->step) {
                animation->state -= animation->step;
            } else {
                animation->direction = ANIMATION_DIRECTION::UP;
                animation->state = 0;
            }
        }
        setAll(_transformPulse(animation->state, animation->color));
```

Reply:
The suggestion changes the logic. A replacement without using LED_PWM_MIN/MAX in commit 2c34e5377b9ad3f4502a7a233d0a77402f6a7c0c.